src/server/worker/api_server.go

```diff
@@ -957,3 +744,420 @@ func (a *APIServer) updateJobState(stm col.STM, jobInfo *pps.JobInfo, state pps.
 	jobs.Put(jobInfo.Job.ID, jobInfo)
 	return nil
 }
+
+type acquireDatumsFunc func(low, high int64) (failedDatumID string, _ error)
+
+func (a *APIServer) acquireDatums(ctx context.Context, jobID string, chunks *Chunks, logger *taggedLogger, process acquireDatumsFunc) error {
+	complete := false
+	for !complete {
+		var low int64
+		var high int64
+		var found bool
+		if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+			found = false
+			locks := a.locks(jobID).ReadWrite(stm)
+			// we set complete to true and then unset it if we find an incomplete chunk
+			complete = true
+			for _, high = range chunks.Chunks {
+				var chunkState ChunkState
+				if err := locks.Get(fmt.Sprint(high), &chunkState); err != nil {
+					if col.IsErrNotFound(err) {
+						found = true
+					} else {
+						return err
+					}
+				}
+				// This gets triggered either if we found a chunk that wasn't
+				// complete or if we didn't find a chunk at all.
+				if chunkState.State == ChunkState_RUNNING {
+					complete = false
+				}
+				if found {
+					break
+				}
+				low = high
+			}
+			if found {
+				return locks.PutTTL(fmt.Sprint(high), &ChunkState{State: ChunkState_RUNNING}, ttl)
```

Comment:
So I was under the impression that if an `STM` failed due to contention it would be retried. Do you happen to know if that's the case? If not I guess we should wrap this in retry logic and perhaps add a randomized delay, if it does retry by itself I think that's sufficient though.